### ISSUE TITLE
asciidoctor-diagram leaves output in project root

### ISSUE THREAD

**Issue by nicorikken** [asciidoctor/asciidoctorj#573]
I am using `asciidoctorj` in combination with `asciidoctorj-diagram` in the [Perun](https://github.com/hashobject/perun) static site generator writen in Clojure. It is based on the [Boot](http://boot-clj.com/) framework, which uses temporary directories called [Filesets](https://github.com/boot-clj/boot/wiki/Filesets) in the process of processing content.

To make sure the images end up on the right target location (in a `tmp-dir`), I set both the `base_dir` in the options and the `outdir` in the attributes to the `tmp-dir`. This works as expected:

- tmp-dir `~/.boot/cache/tmp/home/nico/project-root/h7n/g2jvv7/`
- base_dir `~/.boot/cache/tmp/home/nico/project-root/h7n/g2jvv7/public/posts/`
- outdir `~/.boot/cache/tmp/home/nico/project-root/h7n/g2jvv7/public/posts/`
- image `~/.boot/cache/tmp/home/nico/project-root/h7n/g2jvv7/public/posts/image.svg`

These temporary files get committed and pushed to the `target/` folder. This also works with a relative `imagesdir`. So far so good:

```
$ ls home/nico/project-root/target/public/posts
image.svg
post.html
```

Unfortunately I also end up with a cache file and the image in the root of the project:

```
$ cd home/nico/project-root
$ find . '*.svg*'
./image.svg
./target/public/posts/image.svg
./.asciidoctor/diagram/image.svg.cache
```

Ideally I would like to have it end up in a temporary directory as well, so the root of my project will not be cluttered, as I would make sure that part of the temporary directory is not copied over to the final `target/` directory.

As far as I understand from the Java code, the `base_dir` in the convert options determines the `currentDirectory`: https://github.com/asciidoctor/asciidoctorj/blob/0b76a6aee5429c2c97863a64bd85ae8c1b9c1061/asciidoctorj-core/src/main/java/org/asciidoctor/internal/JRubyAsciidoctor.java#L303-L305
I am certain the `base_dir` was set, as it is required for getting the final output in the right location, and without it it won't work.
Yet `base_dir` is not sufficient to get these temporary files in the right location.

What can I do to get rid of these temporary files? Ideally I would execute the `convert` command in a temporary directory, to prevent any cluttering in the first place. Cleaning these files afterwards might be an alternative solution.

**Comment by abelsromero** [comment]
What you mention about `base_dir` is odd to me. According to [this](http://asciidoctor.org/man/asciidoctor/#security-settings), `base_dir` is used to calculate the root of the sources. For instance, it is used when looking for files with the include directive or as base for `images` if it's not absolute. So it should not affect how files appear in the target path.

> Unfortunately I also end up with a cache file and the image in the root of the project:

Based in my experience with other build tools (mainly maven and a bit of gradle) and assuming that `outdir` gets passed to asciidoctor as the `destination-dir` option, I'd say that you cannot get rid of the additionla files. Not without some code at least. Also, not I am not familiar with Boot :% and how it integrates Asciidoctorj.
`asciidoctor-diagram` honours `destination-dir`+`images` attribute to place the cache and the final images. So if you want to place them somewhere temporarily, you have to set the temp path as `destination_dir`, and then write some code to filter and copy things over.
Given that we have the same concern with the cache dir in the maven-plugin, this is something we discuss to push to asciidoctor-diagram. Adding an attribute to set the cache dir maybe.

As always, happy to be proven wrong.

**Comment by nicorikken** [comment]
If I leave out the `base_dir` it creates a container based from the normal project-root:
```
asciidoctor-diagram: ERROR: Failed to generate image: Start path /home/nico/.boot/cache/tmp/home/nico/project-root/p2k/3v4d88/public/posts/ is outside of jail: /home/nico/project-root (disallowed in safe mode)
```

I did not know of the `destination_dir` option, but now I found it https://github.com/asciidoctor/asciidoctorj/blob/0b76a6aee5429c2c97863a64bd85ae8c1b9c1061/asciidoctorj-core/src/main/java/org/asciidoctor/internal/AsciidoctorUtils.java#L52 I gave it a try, but whether or not I set it (equal to `outdir` and `base_dir`) does not make a difference, so also not for the caching files.

After some more trial and error, I found the solution to my problem!

I was handling the document twice, once to generate the html and produce the images, and second to read out the document attributes. But since I was using the same asciidoctor container with the asciidoctor-diagram library, also in the second instance (reading attributes) the images were generated.
But in the second instance I did not set any options or attributes, so `base_dir` and `to_dir` were not set, and defaulted to the `project-root/`. By using the same context as when converting the source, I got the temporary files to end up in the tmp-dir, which did not clutter the end result.

So certainly my own mistake. I'll close this issue now.
@abelsromero thanks for thinking along!
